File: landscape-webui/src/api/dns_service.ts
import { CheckDnsReq, CheckDnsResult } from "@/rust_bindings/dns";
import api from ".";
import { ServiceStatus } from "@/lib/services";

export async function get_dns_status(): Promise<ServiceStatus> {
  let data = await api.api.get("sys_service/dns");
  // console.log(data.data);
  return new ServiceStatus(data.data);
}

export async function start_dns_service(
  udp_port: number
): Promise<ServiceStatus> {
  let data = await api.api.post("sys_service/dns", {
    udp_port,
  });
  // console.log(data.data);
  return new ServiceStatus(data.data.status);
}

export async function stop_dns_service(): Promise<ServiceStatus> {
  let data = await api.api.delete("sys_service/dns");
  // console.log(data.data);
  return new ServiceStatus(data.data.status);
}

export async function check_domain(req: CheckDnsReq): Promise<CheckDnsResult> {
  let data = await api.api.get("sys_service/dns/check", {
    params: { ...req },
  });
  return data.data;
}

File: landscape-webserver/src/sys_service/dns_service.rs
use axum::{
    extract::{Query, State},
    routing::get,
    Router,
};
use landscape_common::service::DefaultWatchServiceStatus;
use landscape_dns::{CheckChainDnsResult, CheckDnsReq};

use crate::LandscapeApp;

use crate::{api::LandscapeApiResp, error::LandscapeApiResult};

pub async fn get_dns_paths() -> Router<LandscapeApp> {
    Router::new()
        .route("/dns", get(get_dns_service_status).post(start_dns_service).delete(stop_dns_service))
        .route("/dns/check", get(check_domain))
}

async fn get_dns_service_status(
    State(state): State<LandscapeApp>,
) -> LandscapeApiResult<DefaultWatchServiceStatus> {
    LandscapeApiResp::success(state.dns_service.get_status().await)
}

async fn start_dns_service(State(state): State<LandscapeApp>) {
    state.dns_service.start_dns_service().await;
}

async fn stop_dns_service(State(state): State<LandscapeApp>) {
    state.dns_service.stop().await;
}

async fn check_domain(
    State(state): State<LandscapeApp>,
    Query(req): Query<CheckDnsReq>,
) -> LandscapeApiResult<CheckChainDnsResult> {
    LandscapeApiResp::success(state.dns_service.check_domain(req).await)
}
